MAINT: Better error message from histogram2d

jeertmans opened the pull request:
Hello,

I added the error message as proposed by @WarrenWeckesser in #20050.

rossbar commented on numpy/lib/tests/test_twodim_base.py:
```diff
@@ -295,6 +298,17 @@ def __array_function__(self, function, types, args, kwargs):
         r = histogram2d(xy, xy, weights=s_d)
         assert_(r, ((ShouldDispatch,), (xy, xy), dict(weights=s_d)))
 
+    def test_bad_length(self):
+        x_lengths = [10, 30, 100]
+        y_lengths = [11, 50, 150]
+
+        for m, n in zip(x_lengths, y_lengths):
+            with pytest.raises(ValueError, 
+                    match='x and y must have the same length.'):
+                x = np.ones(m)
+                y = np.ones(n)
+                histogram2d(x, y)
+
```
A few other minor suggestions for the test:
 1. As of now, all the test cases are roughly the same, even if the exact values of lengths are different. It'd be more useful to have e.g. `x_lengths = [10, 20]` and `y_lengths = [11, 19]`, because in that case you'd be testing both the `len(x) > len(y)` and `len(x) < len(y)` case. In the (highly unlikely) scenario that someone in the future modified your change to `histogram2d` to be, say, `if len(x) <= len(y):`, then the current test would continue to pass, even though it shouldn't.
 2. A better way to test multiple, different cases with the same test is to use [pytest's parametrize feature](https://docs.pytest.org/en/6.2.x/parametrize.html).

With these suggestions, the test might look something more like:

```suggestion
    pytest.mark.parametrize(("x_len", "y_len"), [(10, 11), (20, 19)])
    def test_bad_length(self, x_len, y_len):
        x, y = np.ones(x_len), np.ones(y_len)
        with pytest.raises(ValueError, match='x and y must have the same length.'):
            histogram2d(x, y)

```

rossbar reviewed (commented):
No major blockers, but since @WarrenWeckesser did such an excellent job reviewing I thought I'd add my two cents!

jeertmans commented:
Well @rossbar I cannot agree more with your suggestion, so I just committed your proposed changes. If there is anything else I can do, please feel free to ask.

jeertmans committed:
BUG: Check that the lengths of the inputs to histogram2d are the same.

Prior to that, the error message would not give much clue about what the real source of the problem was. Closes gh-20050


Update tests for histogram2d

Add pytest parametrization, and provide more robust testing on x and y lengths.

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>
Update test_twodim_base.py

Lint
Linting

rossbar reviewed (approved):
LGTM, thanks @jeertmans !
